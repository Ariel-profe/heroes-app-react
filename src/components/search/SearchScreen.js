import React, { useMemo } from 'react'
import queryString from 'query-string';

import { useLocation } from 'react-router';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = ({history}) => {

    const location = useLocation();
    
    const { q = '' } = queryString.parse( location.search )
    
    const [values, handleInputChange] = useForm({
        searchText: q
    });
    
    const { searchText } = values;
    
    const heroesFilter = useMemo(() => getHeroesByName( q ), [q])
    

    const handleSearch = (e) => {
        e.preventDefault()

        history.push(`?q=${searchText}`);
    };

    return (
        <div>
            <h1> Search screen</h1>
            <hr/>

            <div className="row">

                <div className="col-5 ">
                   <h4> Search form</h4> 
                   <hr/>

                   <form className="d-grid gap-2 " onSubmit={ handleSearch }>
                       <input
                        type="text"
                        placeholder="Find your hero"
                        autoComplete="off"
                        className="form-control d-grid gap-2"
                        name="searchText"
                        value={searchText}
                        onChange={ handleInputChange }
                       />
                       
                       <button
                            type="submit"
                            className="btn m-2 btn-outline-primary"
                       >
                            Seach...
                       </button>
                   </form>
                </div>
                
                <div className="col-7">
                    <h4> Results </h4>
                    <hr/>

                   { 
                        (q === '' ) 
                            && <div className="alert alert-info">
                        Search a hero
                                </div>
                    }

                    { 
                        (q !== '' && heroesFilter.length === 0) 
                            && <div className="alert alert-danger">
                        There is no a hero with { q }
                                </div>
                    }

                    {
                        heroesFilter.map(hero => (
                            <HeroCard 
                                key={hero.id}
                                {...hero}
                            />
                        ))
                    }
                </div>

                    


            </div>


        </div>
    )
}
