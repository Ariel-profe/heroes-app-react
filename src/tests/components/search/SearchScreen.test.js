import { mount } from "enzyme";
import { Route } from "react-router-dom";
import { MemoryRouter } from "react-router-dom/cjs/react-router-dom.min";
import { SearchScreen } from "../../../components/search/SearchScreen";


describe('Pruebas en el <SearchScreen />', () => {

    const wrapper = mount(
        <MemoryRouter initialEntries={ ['/search'] }>
            <Route path={'/search'} component={ SearchScreen } />
        </MemoryRouter>
    );

    test('Debe mostrarse correctamente con valores por defecto', () => {

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').exists()).toBe(true);
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero');
        
    });

    test('Debe mostrar a Batman e input con el valor del queryString', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search?q=batman'] }>
                <Route path={'/search'} component={ SearchScreen } />
            </MemoryRouter>
        );

        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();

    });

    test('Debe mostrar un error si no se encuentra el heroe', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search?q=batman123'] }>
                <Route path={'/search'} component={ SearchScreen } />
            </MemoryRouter>
        );

        expect(wrapper.find('.alert-danger').exists()).toBe(true)
        expect(wrapper.find('.alert-danger').text().trim()).toBe('There is no a hero with batman123');
    });

    test('Debe llamar el push del history', () => {
        
        const historyMock = {
            push: jest.fn()
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search?q=batman123'] }>
                <Route 
                    path={'/search'} 
                    component={ () => <SearchScreen history={historyMock} /> } 
                />
            </MemoryRouter>
        );

        //Hacemos el cambio de la caja de texto
        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'batman'
            }
        });

        //Hacemos el submit del formulario
        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });

        expect(historyMock.push).toHaveBeenCalledWith(`?q=batman`)

    });

});