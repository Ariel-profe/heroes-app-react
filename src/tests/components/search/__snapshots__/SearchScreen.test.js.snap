// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <SearchScreen /> Debe mostrar a Batman e input con el valor del queryString 1`] = `
<div>
  <h1>
     Search screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5 "
    >
      <h4>
         Search form
      </h4>
      <hr />
      <form
        className="d-grid gap-2 "
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control d-grid gap-2"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn m-2 btn-outline-primary"
          type="submit"
        >
          Seach...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
         Results 
      </h4>
      <hr />
      <a
        className="my-card"
        href="/hero/dc-batman"
        onClick={[Function]}
      >
        <img
          alt="Batman"
          className="img img-responsive"
          src="./assets/heroes/dc-batman.jpg"
        />
        <div
          className="profile-name"
        >
          Batman
        </div>
        <div
          className="profile-position"
        >
          Bruce Wayne
        </div>
        <div
          className="profile-overview"
        >
          <div
            className="profile-overview"
          >
            <div
              className="row"
            >
              <div
                className="col-ms-4"
              >
                <h3>
                  DC Comics
                </h3>
                <p>
                  Primera aparici√≥n: 
                  <br />
                  Detective Comics #27
                </p>
              </div>
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>
</div>
`;

exports[`Pruebas en el <SearchScreen /> Debe mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
     Search screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5 "
    >
      <h4>
         Search form
      </h4>
      <hr />
      <form
        className="d-grid gap-2 "
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control d-grid gap-2"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn m-2 btn-outline-primary"
          type="submit"
        >
          Seach...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
         Results 
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
